internal/builder/resourceset.go

```diff
@@ -0,0 +1,205 @@
+	for i, ji := range jsonInputs {
+		inp := make(map[string]any, len(ji))
+		for k, v := range ji {
+			var data any
+			if err := json.Unmarshal(v.Raw, &data); err != nil {
+				return nil, fmt.Errorf("failed to unmarshal inputs[%d]: %w", i, err)
+			}
+			inp[k] = data
+		}
+		inputs = append(inputs, inp)
+	}
+
+	var objects []*unstructured.Unstructured
+
+	// build resources from JSON templates
+	for i, tmpl := range templates {
+		if len(inputs) == 0 {
+			object, err := BuildResource(tmpl, nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resource: %w", err)
+			}
+
+			objects = append(objects, object)
+			continue
+		}
+
+		for _, input := range inputs {
+			object, err := BuildResource(tmpl, input)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resources[%d]: %w", i, err)
+			}
+
+			// exclude object based on annotations
+			if val := object.GetAnnotations()[fluxcdv1.ReconcileAnnotation]; val == fluxcdv1.DisabledValue {
+				continue
+			}
+
+			// deduplicate objects
+			if !containsObject(objects, object) {
+				objects = append(objects, object)
+			}
+		}
+	}
+
+	// build resources from multi-doc YAML template
+	if yamlTemplate != "" {
+		var objectsFromTemplate []*unstructured.Unstructured
+		if len(inputs) == 0 {
+			objs, err := BuildResourcesFromYAML(yamlTemplate, nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resource: %w", err)
+			}
+
+			objectsFromTemplate = append(objectsFromTemplate, objs...)
+		}
+		for _, input := range inputs {
+			objs, err := BuildResourcesFromYAML(yamlTemplate, input)
+			if err != nil {
+				return nil, fmt.Errorf("failed to build resources: %w", err)
+			}
+			objectsFromTemplate = append(objectsFromTemplate, objs...)
+		}
+
+		for _, object := range objectsFromTemplate {
+			// exclude object based on annotations
+			if val := object.GetAnnotations()[fluxcdv1.ReconcileAnnotation]; val == fluxcdv1.DisabledValue {
+				continue
+			}
+			// deduplicate objects
+			if !containsObject(objects, object) {
+				objects = append(objects, object)
+			}
+		}
+	}
+
+	return objects, nil
+}
+
+// BuildResource builds a Kubernetes resource from a JSON template using the provided inputs.
+// Template functions are provided by the slim-sprig library https://go-task.github.io/slim-sprig/.
+// In addition, the slugify function is available to generate slugs from strings using https://github.com/gosimple/slug/.
+func BuildResource(tmpl *apix.JSON, inputs map[string]any) (*unstructured.Unstructured, error) {
+	yamlTemplate, err := yaml.JSONToYAML(tmpl.Raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert template to YAML: %w", err)
+	}
+
+	tp, err := template.New("res").
+		Delims("<<", ">>").
+		Funcs(sprig.HermeticTxtFuncMap()).
+		Funcs(template.FuncMap{"slugify": slug.Make}).
+		Funcs(template.FuncMap{"inputs": func() any { return inputs }}).
+		Funcs(template.FuncMap{"toYaml": toYaml, "mustToYaml": mustToYaml}).
```

Comment:
In order to keep the templating engine consistent, let's extract this part to a function that takes `inputs` as argument and call it in both `BuildResource` and `BuildResourcesFromYAML`.